### ISSUE TITLE
SAM3SemanticPredictor high mask memory use

### ISSUE THREAD

**Issue by BSchilperoort** [ultralytics/ultralytics#26194]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Predict

### Bug

The SAM3SemanticPredictor upscales all kept masks to the original image size as a single float32 tensor before thresholding them, which costs 4 bytes per pixel per instance. On large images with many instances this can cause memory issues.

In our project we [subclassed the predictor, and threshold the masks N at a time, thresholding them in batches](https://github.com/Urban-M4/streetscapes/blob/bb4e825533697008c8af961ae734e994f281f6db/src/streetscapes/models/sam3/predictor.py). This reduces memory usage while not visibly impacting performance.

### Environment

Ultralytics 8.4. Other info not relevant.

### Minimal Reproducible Example

Problem is evident from looking at the source code (line 2361);
https://github.com/ultralytics/ultralytics/blob/fa39b69cb3a48303efe5c1ec3b271e97352d7d71/ultralytics/models/sam/predict.py#L2356-L2366

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @BSchilperoort, thank you for reporting the high memory use in `SAM3SemanticPredictor` and for providing a batching-based subclass implementation. Thresholding masks in batches before upscaling appears to be a useful improvement for large images with many instances, and we welcome a PR with focused tests and a memory/performance comparison.

If possible, please include the representative image dimensions, instance count, batch size, and peak memory measurements in the PR so the impact can be evaluated consistently. This is an automated response, and an Ultralytics engineer will assist soon.

**Comment by BSchilperoort** [comment]
> include the representative image dimensions, instance count, batch size, and peak memory measurements

Images in our project are often 12K panoramas, i.e. ~72MP. Instances vary between 5-100 depending on prompts (identifying all windows of a building facade).
In the extreme case; a 12K panorama with 100 instances this would mean the float32 masks alone take up 30GB RAM on GPU.

Batching these in groups of e.g. 8 reduces memory use of float32 masks to 2.4 GB

**Comment by rahultechenable** [comment]
The helper itself looks right, and covering `inference_features` and
`SAM3VideoSemanticPredictor.postprocess` alongside the reported call site is a nice catch.

One thing that may be worth addressing before merge: the chunked path can't be
regression-tested at any reasonable size as written.

```python
chunk = max(1, 2**29 // (shape[0] * shape[1]))

Getting chunk < N requires H*W > 2**29 / N — for even 8 masks that's ~16384x16384,
whose bool output alone is ~2.1 GB. So any CI-sized shape collapses to a single chunk and a
test would silently exercise the old single-shot path. That's what the 11% patch coverage
here reflects, and there are currently no SAM3 tests in the suite at all.

If the budget were hoisted to a module-level constant, a test could shrink it and assert the
chunked result is identical to the single-shot one — mirroring test_process_mask_native_chunked,
which covers the same chunking idea in ops.process_mask_native:

def test_sam3_upscale_masks_chunked(monkeypatch):
    """Chunked SAM3 mask upscaling is identical to upscaling all masks at once."""
    from types import SimpleNamespace

    from ultralytics.models.sam import predict as sam_predict

    torch.manual_seed(0)
    masks, shape = torch.randn(7, 16, 16), (64, 64)
    predictor = SimpleNamespace(model=SimpleNamespace(mask_threshold=0.0))
    ref = torch.nn.functional.interpolate(masks[None].float(), shape, mode="bilinear")[0] > 0.0
    monkeypatch.setattr(sam_predict, "MAX_UPSCALE_ELEMENTS", 2 * 64 * 64)  # force chunks of 2
    out = sam_predict.SAM3SemanticPredictor._upscale_masks(predictor, masks, shape)
    assert out.dtype == torch.bool
    assert torch.equal(out, ref)

No weights and no GPU, so it runs in normal CI. Happy to open a follow-up PR with the
constant plus this test if you'd rather keep this one focused.

```

Two smaller notes:

- out[i : i + chunk] = (F.interpolate(...) > threshold) materializes a chunk-sized bool
  temporary before copying into the view. torch.gt(F.interpolate(...)[0], self.model.mask_threshold, out=out[i : i + chunk]) writes straight into the slice — ~500 MB saved at 12000x6000 with
  chunk=7, small next to the 2 GB float buffer but free.
- ops.process_mask_native budgets 32M pixels (~128 MB float32) for the same upscale-then-
  threshold pattern, versus 2 GiB here. Is the 16x difference intentional, or worth aligning?
